File: DFS/graph/q79_word_search.py
from typing import List


class Solution:
    # 方向数组，相对于当前位置的 4 个方向-横、纵坐标的偏移量，这是一个常见的技巧
    dirs = [(1, 0), (0, 1), (-1, 0), (0, -1)]

    def exist(self, board: List[List[str]], word: str) -> bool:
        if not board: return False
        rows, cols = len(board), len(board[0])
        # rows和cols勿写反！！❤ 且 不可写成 [[False] * cols] * rows【错×】
        visited = [[False] * cols for _ in range(rows)]  # ↑ 引用，某一行改了，所有行会跟着一起改变
        # 遍历起点（棋盘for*2） + 遍历方向（dfs）
        for i in range(rows):
            for j in range(cols):
                visited[i][j] = True
                if self.dfs(board, word, i, j, 0, visited):
                    return True
                visited[i][j] = False
        return False

    def dfs(self, board, word, i, j, idx, visited):
        ch = board[i][j]
        visited[i][j] = True
        if ch != word[idx]: return False
        if idx >= len(word) - 1: return True
        # dfs中遍历方向
        for dir in self.dirs:
            new_x, new_y = i + dir[0], j + dir[1]
            if not self._isBounded(new_x, new_y, board):
                continue
            if visited[new_x][new_y]: continue

            visited[new_x][new_y] = True
            if self.dfs(board, word, new_x, new_y, idx + 1, visited):
                return True
            visited[new_x][new_y] = False
        return False

    def _isBounded(self, new_x, new_y, board):
        m, n = len(board) - 1, len(board[0]) - 1
        return 0 <= new_x <= m and 0 <= new_y <= n


if __name__ == "__main__":
    sol = Solution()
    board = [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]]
    word = "ABCCED"
    res = sol.exist(board, word)
    print(res)

File: Recursion/graph/q79_word_search.java
package Recursion.graph;

public class q79_word_search {
    int m, n;
    char[][] board;
    char[] words;
    int[][] dirs = {{1,0}, {0,1}, {-1, 0}, {0, -1}}; // 四个方向
    public boolean exist(char[][] board, String word) {
        m = board.length; n = board[0].length;
        this.board = board; words = word.toCharArray();
        boolean[][] visited = new boolean[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                // if (!visited[i][j]) {
                visited[i][j] = true;
                if (dfs(0, i, j, visited)) {
                    return true;
                }
                visited[i][j] = false;
                // }
            }
        }
        return false;
    }

    private boolean dfs(int idx, int i, int j, boolean[][] visited) {
        if (words[idx] != board[i][j]) return false;// 条件1：大前提！放在最首！
        if (idx >= words.length-1) return true; // 【易错】：必须-1！且必须在条件1后面！
        visited[i][j] = true;
        for (int[] dir: dirs) {
            int new_i = i + dir[0];
            int new_j = j + dir[1];
            if (check(new_i, new_j, visited)) {
                visited[new_i][new_j] = true;
                if (dfs(idx+1, new_i, new_j, visited)) {
                    return true;
                }
                visited[new_i][new_j] = false;
            }
        }
        return false;
    }

    private boolean check(int i, int j, boolean[][] visited) {
        return 0 <= i && i < m && 0 <= j && j < n && !visited[i][j];
    }


    // 写法2 - isFound 全局变量
//    int m, n;
//    char[][] board;
//    char[] words;
    int[] _x = {1, 0, -1, 0};
    int[] _y = {0, 1, 0, -1};
    int len;
    boolean[][] visited;
    boolean isFound = false;
    public boolean exist2(char[][] board, String word) {
        m = board.length; n = board[0].length; len = word.length();
        this.visited = new boolean[m][n];
        this.board = board; words = word.toCharArray();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (visited[i][j]) continue;
                if (board[i][j] == words[0]) {
                    dfs0(0, i, j, visited);
                }
            }
        }
        return isFound;
    }

    private void dfs0(int idx, int i, int j, boolean[][] visited) {
        if (isFound) return;
        if (words[idx] != board[i][j]) return;
        if (idx >= len-1) { // ?判断条件！
            isFound = true;
            return;
        }

        visited[i][j] = true;

        for (int dir = 0; dir < 4; dir++) {
            int newX = i + _x[dir], newY = j + _y[dir];
            if (check(newX, newY)) {
                dfs0(idx+1, newX, newY, visited);
            }
        }

        visited[i][j] = false;
    }

    private boolean check(int i, int j) {
        return 0 <= i && i < m && 0 <= j && j < n && !visited[i][j];
    }
}
